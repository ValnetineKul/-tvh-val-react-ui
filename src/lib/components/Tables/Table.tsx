import type { FC } from 'react';
import React, { useState, useEffect, useRef, createContext } from 'react';
import { default as MuiTable } from '@mui/material/Table';
import TableContainer from './TableContainer';
import useStyles from './Table.styles';
import type { TableShadows } from './Table.types';

export interface TableProps {
  disabledStickyHeader?: boolean;
  tableContainerClassName?: string;
  className?: string;
  children: React.ReactNode;
}

const shadows: TableShadows = {
  shouldShowHeaderShadow: false,
  shouldShowVerticalShadowStart: false,
  shouldShowVerticalShadowEnd: false,
  rightShadowOffset: 0,
};

export const ShadowsContext = createContext(shadows);

const Table: FC<TableProps> = ({ disabledStickyHeader = false, tableContainerClassName, className, children }) => {
  const [shadowsShown, setShadowsShown] = useState(shadows);
  const [tableWidth, setTableWidth] = useState(0);
  const [tableHeight, setTableHeight] = useState(0);
  const { classes, cx } = useStyles({
    rightShadowOffset: tableWidth,
    verticalOffset: tableHeight,
  });
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const onScroll = () => {
      if (!ref.current) {
        return;
      }
      const { scrollTop, scrollWidth, scrollLeft, clientWidth, clientHeight } = ref.current;
      setShadowsShown({
        shouldShowHeaderShadow: scrollTop > 0,
        shouldShowVerticalShadowStart: scrollLeft > 0,
        shouldShowVerticalShadowEnd: Math.ceil(scrollLeft) + clientWidth < scrollWidth,
        rightShadowOffset: clientWidth,
      });
      setTableWidth(clientWidth);
      setTableHeight(clientHeight);
    };
    onScroll();
    const node = ref.current;
    if (node) {
      node.addEventListener('scroll', onScroll);
    }
    window.addEventListener('resize', onScroll);
    return () => {
      if (node) {
        node.removeEventListener('scroll', onScroll);
      }
      window.removeEventListener('resize', onScroll);
    };
  }, []);

  const renderShadows = () => {
    return (
      <>
        {shadowsShown.shouldShowHeaderShadow && (
          <div className={cx(classes.headerShadow, { [classes.disabledStickyHeaderShadow]: disabledStickyHeader })} />
        )}
        {shadowsShown.shouldShowVerticalShadowStart && (
          <div className={cx(classes.verticalShadow, classes.startShadow)} />
        )}
        {shadowsShown.shouldShowVerticalShadowEnd && <div className={cx(classes.verticalShadow, classes.endShadow)} />}
      </>
    );
  };

  return (
    <ShadowsContext.Provider value={shadowsShown}>
      <TableContainer className={tableContainerClassName} ref={ref}>
        {renderShadows()}
        <MuiTable stickyHeader={!disabledStickyHeader} className={cx(className, classes.root)}>
          {children}
        </MuiTable>
      </TableContainer>
    </ShadowsContext.Provider>
  );
};

export default Table;
