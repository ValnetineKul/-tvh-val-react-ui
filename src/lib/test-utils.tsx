import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
