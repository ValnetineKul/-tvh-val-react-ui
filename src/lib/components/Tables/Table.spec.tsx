import React from 'react';
import { render } from '../../test-utils';
import Table from './Table';

describe('Table', () => {
  it('Should render correctly', () => {
    const { container } = render(
      <Table>
        <tbody>
          <tr>
            <td>Children</td>
          </tr>
        </tbody>
      </Table>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
