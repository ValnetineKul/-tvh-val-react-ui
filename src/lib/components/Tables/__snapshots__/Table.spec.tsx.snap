// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Should render correctly 1`] = `
<div
  class="MuiTableContainer-root mui-rorn0c-MuiTableContainer-root tss-1giqlv2-Surface-root-TableContainer-root"
>
  <table
    class="MuiTable-root MuiTable-stickyHeader tss-12i0kxp-Table-root mui-1u90bpa-MuiTable-root"
  >
    <tbody>
      <tr>
        <td>
          Children
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
